pkg/bundle/sync.go

```diff
@@ -165,24 +174,61 @@ func encodeJKS(trustBundle string) ([]byte, error) {
 
 		c, err := x509.ParseCertificate(p.Bytes)
 		if err != nil {
-			return []byte{}, err
+			return nil, fmt.Errorf("got invalid cert when trying to encode JKS: %w", err)
 		}
 
-		ks.SetTrustedCertificateEntry(c.Issuer.String(), jks.TrustedCertificateEntry{
+		// Note on CreationTime:
+		// Debian's JKS trust store sets the creation time to match the time that certs are added to the
+		// trust store (i.e., it's effectively time.Now() at the instant the file is generated).
+		// Using that method would make our JKS files in trust-manager non-deterministic, leaving us with
+		// two options if we want to maintain determinism:
+		// - Using something from the cert being added (e.g. NotBefore / NotAfter)
+		// - Using a fixed time (i.e. unix epoch)
+		// We use NotBefore here, arbitrarily.
+
+		alias := jksAlias(c)
+
+		err = ks.SetTrustedCertificateEntry(alias, jks.TrustedCertificateEntry{
 			CreationTime: c.NotBefore,
 			Certificate: jks.Certificate{
 				Type:    "X509",
 				Content: p.Bytes,
 			},
 		})
+
+		if err != nil {
+			// this error should never happen if we set jks.Certificate correctly
+			return nil, fmt.Errorf("failed to add cert with alias %q to trust store: %w", alias, err)
+		}
 	}
 
 	buf := &bytes.Buffer{}
-	ks.Store(buf, []byte(defaultJKSPassword))
+
+	err := ks.Store(buf, password)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create JKS file: %w", err)
+	}
 
 	return buf.Bytes(), nil
 }
 
+// jksAlias creates a JKS-safe alias for the given certificate, such that
+// any two certificates will have a different aliases unless they're
+// identical in every way.
+// This unique alias fixes an issue where we used the Issuer field as an
+// alias, leading to different certs being treated as identical.
+func jksAlias(cert *x509.Certificate) string {
+	certHashBytes := sha256.Sum256(cert.Raw)
```

Comment:
Can we do a panic in case cert.Raw is nil/ empty? I assume `jksAlias` only works for `*x509.Certificate` values that are returned by `x509.ParseCertificate`?

Reply:
@SgtCoDFish

Reply:
Please don't panic in kube controllers. Bubble up the error that this is an irreconcilable state, and update the crd object with this state + condition.

Reply:
Makes sense to be defensive like that yeah :+1: